File: 1-two-sum.py
class Solution:

    def two_sum(self, nums: list, target: int) -> list | None:
        for x in range(0, len(nums) - 1):
            for y in range(x + 1, len(nums)):
                if nums[x] == nums[y]:
                    return [x, y]
    
if __name__ == '__main__':
    sol = Solution()
    nums = [0, 4, 3, 0]
    print(sol.two_sum(nums, 0))

File: 1-two-sum-go/1-two-sum.go
package main

func main() {
	twoSumBrute([]int{1, 2, 3, 4}, 5)
	// var results []int = twoSum([]int{1, 2, 3, 4}, 5)
	// for i := range results {
	// 	println(results[i])
	// }
	twoSumMap([]int{1, 2, 3, 4}, 5)
}

func twoSumBrute(nums []int, target int) []int {
	var results = []int{}

	for i := range nums {
		for j := i + 1; j < len(nums); j++ {
			if nums[i]+nums[j] == target {
				results = append(results, i, j)
				break
			}
		}

		if len(results) > 0 {
			break
		}
	}

	return results
}

func twoSumMap(nums []int, target int) []int {
	var results = []int{}
	var differences = make(map[int]int)

	for i, value := range nums {
		if index, ok := differences[value]; ok {
			results = append(results, index, i)
			break
		}

		differences[target-value] = i
	}

	return results
}
